src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1971 @@
+                        override this.ProcessNext (input:'T) : bool =
+                            this.State.Add (keyf input) && TailCall.avoid (next.ProcessNext input) }}
+
+        [<CompiledName "Max">]
+        let inline max (source:ISeq<'T>) : 'T when 'T:comparison =
+            source.Fold (fun _ ->
+                upcast { new FolderWithPostProcessing<'T,'T,bool>(Unchecked.defaultof<'T>,true) with
+                    override this.ProcessNext value =
+                        if this.State then
+                            this.State <- false
+                            this.Result <- value
+                        elif value > this.Result then
+                            this.Result <- value
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ =
+                        if this.State then
+                            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                    override this.OnDispose () = () })
+
+        [<CompiledName "MaxBy">]
+        let inline maxBy (f:'T->'U) (source:ISeq<'T>) : 'T when 'U:comparison =
+            source.Fold (fun _ ->
+                upcast { new FolderWithPostProcessing<'T,'T,Values<bool,'U>>(Unchecked.defaultof<'T>,Values<_,_>(true,Unchecked.defaultof<'U>)) with
+                    override this.ProcessNext value =
+                        match this.State._1, f value with
+                        | true, valueU ->
+                            this.State._1 <- false
+                            this.State._2 <- valueU
+                            this.Result <- value
+                        | false, valueU when valueU > this.State._2 ->
+                            this.State._2 <- valueU
+                            this.Result <- value
+                        | _ -> ()
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ =
+                        if this.State._1 then
+                            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                    override this.OnDispose () = () })
+
+        [<CompiledName "Min">]
+        let inline min (source:ISeq<'T>) : 'T when 'T:comparison =
+            source.Fold (fun _ ->
+                upcast { new FolderWithPostProcessing<'T,'T,bool>(Unchecked.defaultof<'T>,true) with
+                    override this.ProcessNext value =
+                        if this.State then
+                            this.State <- false
+                            this.Result <- value
+                        elif value < this.Result then
+                            this.Result <- value
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ =
+                        if this.State then
+                      
```

Comment:
Well the "I" in ISeq was stands for inline :-)

So originally these were inline as ISeq was an internal type that I was just using as a helper for my replacement of Seq, so they were inlined deliberately so that they would then inhabit the Seq module, but after testing performance version Nessos's Streams, I found that the only reason ISeq was slower was because it wasn't inlining functions (as it was being accessed through Seq) so I made the ISeq interface public and that's where we are.

So if we do the next phase of this project, which is to gut Seq and replace it as stub calls to ISeq 
(as discuessed [here](https://github.com/Microsoft/visualfsharp/pull/2587#issuecomment-289181712)) then I think the overkill of inline in ISeq is fine. Most users not care about ISeq; they will get the majority of the performance boost without touching a line of code.

For those that do want to squeeze the last bit out of their piplelines the can start passing ISeqs around and they get inlined code...